File: Assets/_Project/Scripts/Rendering/HexTile.cs
using UnityEngine;
using HexBattle.Core;

namespace HexBattle.Rendering
{
    /// <summary>
    /// Component attached to every hex tile GameObject in the scene.
    /// Handles per-tile visual state: highlights, terrain overlays, terrain VFX.
    /// </summary>
    [RequireComponent(typeof(MeshRenderer))]
    public class HexTile : MonoBehaviour
    {
        // ── Coords ────────────────────────────────────────────────────────────
        public Vector2Int Coords;        // axial (q, r)
        public TerrainType CurrentTerrain { get; private set; } = TerrainType.None;
        public int TurnsLeft  { get; private set; }
        public int Integrity  { get; private set; }    // for rock piles etc.

        // ── Materials ─────────────────────────────────────────────────────────
        [Header("Materials")]
        public Material BaseMaterial;
        public Material MoveHighlightMat;    // blue emissive
        public Material AttackHighlightMat;  // red emissive
        public Material PathHighlightMat;    // white emissive
        public Material AoeMat;             // orange emissive

        // ── Terrain VFX prefabs ───────────────────────────────────────────────
        [Header("Terrain VFX Prefabs")]
        public GameObject BurnZoneVfx;
        public GameObject IceZoneVfx;
        public GameObject PoisonTrapVfx;
        public GameObject ResonanceVfx;
        public GameObject PeishVfx;
        public GameObject MistVfx;
        public GameObject SlowZoneVfx;

        private MeshRenderer _renderer;
        private GameObject   _activeTerrainVfx;

        private void Awake()
        {
            _renderer = GetComponent<MeshRenderer>();
        }

        // ── Highlights ────────────────────────────────────────────────────────
        public void SetHighlight(HighlightType type)
        {
            _renderer.material = type switch
            {
                HighlightType.MoveRange   => MoveHighlightMat,
                HighlightType.AttackRange => AttackHighlightMat,
                HighlightType.Path        => PathHighlightMat,
                HighlightType.AoeRange    => AoeMat,
                _                        => BaseMaterial
            };
        }

        public void ClearHighlight() => SetHighlight(HighlightType.None);

        // ── Terrain ───────────────────────────────────────────────────────────
        public void SetTerrain(TerrainType terrain, int turns = -1)
        {
            CurrentTerrain = terrain;
            TurnsLeft      = turns;

            // Destroy previous VFX
            if (_activeTerrainVfx != null)
                Destroy(_activeTerrainVfx);

            // Spawn correct VFX
            GameObject prefab = terrain switch
            {
                TerrainType.BurnZone      => BurnZoneVfx,
                TerrainType.IceZone       => IceZoneVfx,
                TerrainType.PoisonTrap    => PoisonTrapVfx,
                TerrainType.ResonanceZone => ResonanceVfx,
                TerrainType.PeishZone     => PeishVfx,
                TerrainType.MistZone      => MistVfx,
                TerrainType.SlowZone      => SlowZoneVfx,
                _ => null
            };

            if (prefab != null)
                _activeTerrainVfx = Instantiate(prefab, transform.position,
                    Quaternion.identity, transform);

            // Update base material tint
            if (BaseMaterial != null)
            {
                Color tint = GetTerrainTint(terrain);
                _renderer.material.SetColor("_BaseColor", tint);
            }
        }

        public void ClearTerrain()
        {
            CurrentTerrain = TerrainType.None;
            TurnsLeft      = 0;
            if (_activeTerrainVfx != null) Destroy(_activeTerrainVfx);
        }

        // ── Tooltip data ──────────────────────────────────────────────────────
        public string GetTooltipTitle()
            => $"Hex ({Coords.x}, {Coords.y})";

        public string GetTooltipBody()
        {
            string terrain = CurrentTerrain == TerrainType.None
                ? "Normal"
                : CurrentTerrain.ToString();
            string turns   = TurnsLeft > 0 ? $" ({TurnsLeft}t left)" : "";
            return $"Terrain: {terrain}{turns}\nStep cost: {GetStepCost()}";
        }

        public int GetStepCost()
            => CurrentTerrain == TerrainType.SlowZone ? 2 : 1;

        // ── Helpers ───────────────────────────────────────────────────────────
        private static Color GetTerrainTint(TerrainType t) => t switch
        {
            TerrainType.BurnZone      => new Color(1f, 0.4f, 0.1f, 1f),
            TerrainType.IceZone       => new Color(0.5f, 0.9f, 1f,   1f),
            TerrainType.PoisonTrap    => new Color(0.7f, 0.2f, 0.8f, 1f),
            TerrainType.ResonanceZone => new Color(0.2f, 0.8f, 0.9f, 1f),
            TerrainType.PeishZone     => new Color(0.5f, 0f,   0.7f, 1f),
            TerrainType.MistZone      => new Color(0.8f, 0.9f, 1f,   1f),
            TerrainType.SlowZone      => new Color(0.5f, 0.4f, 0.2f, 1f),
            TerrainType.SunnyZone     => new Color(1f,   0.9f, 0.3f, 1f),
            TerrainType.RainZone      => new Color(0.2f, 0.5f, 1f,   1f),
            _                         => Color.white,
        };
    }
}

File: src/rendering/HexTile.ts
import { Mesh, StandardMaterial, Color3 } from '@babylonjs/core';
import { HighlightType, TerrainType } from '../core/Enums';

const HIGHLIGHT_COLORS: Record<HighlightType, Color3> = {
  [HighlightType.None]: new Color3(0, 0, 0),
  [HighlightType.MoveRange]: new Color3(0.1, 0.3, 0.9),
  [HighlightType.AttackRange]: new Color3(0.9, 0.1, 0.1),
  [HighlightType.Path]: new Color3(0.9, 0.9, 0.9),
  [HighlightType.AoE]: new Color3(0.9, 0.5, 0.1),
};

const TERRAIN_COLORS: Partial<Record<TerrainType, Color3>> = {
  [TerrainType.BurnZone]: new Color3(0.6, 0.1, 0),
  [TerrainType.SlowZone]: new Color3(0.3, 0, 0.5),
  [TerrainType.PoisonTrap]: new Color3(0.4, 0, 0.4),
  [TerrainType.IceZone]: new Color3(0.4, 0.7, 0.9),
  [TerrainType.ElectricZone]: new Color3(0.8, 0.8, 0),
  [TerrainType.ResonanceField]: new Color3(0.5, 0.2, 0.8),
  [TerrainType.PerishTrap]: new Color3(0.1, 0.1, 0.1),
  [TerrainType.MistVeil]: new Color3(0.7, 0.85, 0.9),
  [TerrainType.SunnyZone]: new Color3(0.9, 0.8, 0.1),
  [TerrainType.RainZone]: new Color3(0.2, 0.4, 0.8),
  [TerrainType.PsychicZone]: new Color3(0.8, 0.3, 0.7),
  [TerrainType.SteelZone]: new Color3(0.5, 0.5, 0.6),
};

export class HexTile {
  coords: { x: number; y: number };
  mesh: Mesh;
  currentTerrain: TerrainType = TerrainType.None;
  highlight: HighlightType = HighlightType.None;
  private mat: StandardMaterial;
  private baseColor: Color3;
  private terrainTurns: number = 0;

  constructor(coords: { x: number; y: number }, mesh: Mesh, mat: StandardMaterial, baseColor: Color3) {
    this.coords = coords;
    this.mesh = mesh;
    this.mat = mat;
    this.baseColor = baseColor;
    this.mesh.metadata = { tile: this };
  }

  setHighlight(type: HighlightType): void {
    this.highlight = type;
    this._updateMaterial();
  }

  setTerrain(type: TerrainType, turns: number): void {
    this.currentTerrain = type;
    this.terrainTurns = turns;
    this._updateMaterial();
  }

  clearTerrain(): void {
    this.currentTerrain = TerrainType.None;
    this.terrainTurns = 0;
    this._updateMaterial();
  }

  private _updateMaterial(): void {
    const terrainColor = TERRAIN_COLORS[this.currentTerrain];
    const diffuse = terrainColor ?? this.baseColor;
    this.mat.diffuseColor = diffuse;

    if (this.highlight !== HighlightType.None) {
      this.mat.emissiveColor = HIGHLIGHT_COLORS[this.highlight];
    } else {
      this.mat.emissiveColor = terrainColor
        ? terrainColor.scale(0.3)
        : new Color3(0, 0, 0);
    }
  }

  getTooltipText(): string {
    let text = `Tile (${this.coords.x},${this.coords.y})`;
    if (this.currentTerrain !== TerrainType.None) {
      text += `\n${TerrainType[this.currentTerrain]} (${this.terrainTurns} turns)`;
    }
    return text;
  }
}
